aten/src/ATen/UndefinedTensor.h

```diff
@@ -6,15 +6,15 @@ namespace at {
 
 struct AT_API UndefinedTensor final : public TensorImpl {
 public:
-  static inline UndefinedTensor * singleton() {
+  static constexpr inline TensorImpl * singleton() {
```

Comment:
hmm...are you sure you want to use UndefinedTensor in this way?  Did you do it because it makes writing a few functions easier in TensorBase?

Reply:
I'm not sure what you mean by this question. I guess I can try restating the justification: if you `s/TensorImpl/UndefinedTensor/`, a static assert in `c10::intrusive_ptr` fails, because when you pass UndefinedTensor as the NullType to `c10::intrusive_ptr<TensorImpl, UndefinedTensor>`, it checks that `UndefinedTensor::singleton()` returns a `TensorImpl*` (NOT the refined type.)

As it turns out, no one really needs the refined `UndefinedTensor` type (because it's purpose in life is to serve as an implementation of the `TensorImpl` interface), so this change didn't have any other consequences.

Reply:
Sorry for not being clear here: what I meant was "are you sure you want to use UndefinedTensor as the NullType singleton", not "are you sure you want to return something of type UndefinedTensor."  We discussed the former and decided to do it to avoid increfs/decrefs.